### ISSUE TITLE
SAT_ORACLE gives wrong answers

### ISSUE THREAD

**Issue by myreen** [HOL-Theorem-Prover/HOL#1170]
This is correct behaviour:
```
> minisatProve.SAT_PROVE ``b ==> T``;
val it = ⊢ b ⇒ T: thm
```
This is incorrect behaviour:
```
> minisatProve.SAT_ORACLE ``b ==> T``;
Exception- SAT_cex ⊢ b ⇒ ¬(b ⇒ T) raised
```
In other words, the oracle version of the interface to SAT returns a bogus theorem for this simple input.

The following shows that one can prove false from the bogus theorem.
```
> (minisatProve.SAT_ORACLE ``b ==> T``; fail()) handle minisatProve.SAT_cex th => th;
val it = ⊢ b ⇒ ¬(b ⇒ T): thm
> it |> GEN_ALL |> SPEC T |> REWRITE_RULE [];
val it = ⊢ F: thm
> Tag.dest_tag (tag it);
val it = (["DISK_THM", "HolSatLib"], []): string list * string list
```
The last line shows that the proof of false depends on a trusted oracle call made by `HolSatLib`. That is to say: this isn't a soundness bug in HOL but rather a bug report about `minisatProve.SAT_ORACLE`.

**Comment by binghe** [comment]
I think the issue is (at least) in the function [invoke_solver" (`src/HolSat/minisatProve.sml`, line 57-78)](https://github.com/HOL-Theorem-Prover/HOL/blob/3f6e78258b82149b95ab354e180b95cd094ec4e7/src/HolSat/minisatProve.sml#L57)

```
fun invoke_solver solver lfn ntm clauseth cnfv vc is_proved svm sva in_name =
    let val nr = Array.length clauseth
    in
      if access(getSolverExe solver,[A_EXEC]) then
        let
          val answer = invokeSat solver T (SOME vc) (SOME nr) svm sva in_name
        in case answer of
              SOME model => (* returns |- model ==> ~t *)
              let val model' = transform_model cnfv lfn model
              in if is_proved then satCheck model' ntm
                 else mk_sat_oracle_thm([],mk_imp(list_mk_conj model',ntm)) end
            | NONE    => (* returns ~t |- F *)
                replay_proof is_proved sva nr in_name solver vc clauseth
                             lfn ntm NONE
        end
      else (* do not have execute access to solver binary, or it doesn't exist*)
        (if nr > !sat_limit then
           warn "SAT solver not found. Using slow internal prover."
         else ();
         DPLL_TAUT (dest_neg ntm))
    end
end
```
In the above function, `is_proved` is false when `SAT_ORACLE` is used, and true when `SAT_PROVE` is used.

Suppose the original SAT problem is “?x. t” (tm), whose negation is “!x. ~t” (ntm). The "answer" (model) (returned by `invokeSat solver T (SOME vc) (SOME nr) svm sva in_name`) here is the first value assignment making “t” true, or making “~t” false; In case that the original problem is indeed SAT, “!x. ~t” must be false, and the "answer" (model) is the value assignments actually making “~t” false.

To verify this "answer" and generate a valid theorem without using oracle, `satCheck model' ntm`.  I think what's actually being checked is the term `model /\ ~t` which should be equivalent to “F”.  When `SAT_ORACLE` is used instead, this `satCheck` call doesn't happen, and instead an equivalent oracle theorem `model ==> ~t` is returned (read the comment "returns |- model ==> ~t" at line 64). I think it should make another oracle theorem `model /\ ~t` instead.

For example, when the input problem is `b ==> T`. The only found model is `b` (meaning `b = T`). But `b ==> ~(b ==>T)` (`model ==> ~t`) is NOT equivalent to `F` as the following code shows:
```
> REWRITE_CONV [] ``b ==> ~(b ==> T)``;
val it = [oracles: DISK_THM] [axioms: ] [] |- b ==> ~(b ==> T) <=> ~b: thm
```
while `model /\ ~t` is indeed equivalent to F:
```
> REWRITE_CONV [] ``b /\ ~(b ==> T)``;
val it = [oracles: DISK_THM] [axioms: ] [] |- b /\ ~(b ==> T) <=> F: thm
```

**Comment by binghe** [comment]
Hmm... I think just replace `mk_sat_oracle_thm([],mk_imp(list_mk_conj model',ntm))` with `mk_sat_oracle_thm([ntm],F)` will fix the bug.   The above mentioned `satCheck()` actually should always return [.] |- F, which after `DISCH_ALL` becomes `[] |- ~t ==> F` (t is the original problem).  If we fully trust the SAT solver and do not do this further verification, `mk_sat_oracle_thm([ntm],F)` will give exactly the same output, only this is a theorem without proof.

**Comment by binghe** [comment]
I think... it's possible that MiniSAT's behavior has changed, say, under new C++ compilers. Previously on Linux ARM64, MiniSAT every wrongly reported SAT for an UNSAT problem, and such a mistake was later captured by HOL when trying to construct the theroem. Now, perhaps MiniSAT is doing the opposite thing: reporting UNSAT for an SAT problem (e.g. ``b ==> T``) but this time the further verification (when constructing the output theorem) has automatically *fixed* the answer thus the issue is hidden. Using `SAT_ORACLE` merely exposed this bug.

**Comment by binghe** [comment]
Now I think it's MiniSAT's problem on certain inputs (`~(b ==> T)` is just one of them). Picking another tautology term like `b \/ ~b` the `SAT_ORACLE` returns correct answer:
```
> SAT_ORACLE ``b \/ ~b``;
val it = [oracles: DISK_THM, HolSatLib] [axioms: ] [] ⊢ b ∨ ¬b: thm
> SAT_ORACLE ``b ==> T``;
Exception- SAT_cex [oracles: HolSatLib] [axioms: ] [] ⊢ b ⇒ ¬(b ⇒ T) raised
```
I tried HOL builds back to k6, and also tried latest HOL built with very old C++ compilers (on Mac OS X 10.6, GCC 4.2), the reported issue with `b ==> T` is always there. Thus now I think it should be MiniSAT 1.x's original bug.

**Comment by mn200** [comment]
We're using such an old version of MiniSAT that no-one other than us will ever fix it.  Is there an obvious test-case we can feed into minisat to demonstrate the problem as a first step?

**Comment by binghe** [comment]
> We're using such an old version of MiniSAT that no-one other than us will ever fix it. Is there an obvious test-case we can feed into minisat to demonstrate the problem as a first step?

Yes, I should first confirm if MiniSAT behaves correct on raw inputs (of DIMACS). But then I found MiniSAT is innocent this time!

In `minisatProve.sml`, if I manually evaluate the ML code (first, the `open`s, then all code inside the `local` block). In function `GEN_SAT`, I got the following values of its local variables:
```
val cnfv = SOME "SP": string option
val in_name = "/var/tmp//MLTEMPECqLFC.cnf": string
val lfn = <Redblackmap(1)>: (term, term) RBM.dict
val ntm = ``~(b ==> T)``: term
val sva = fromList[``T``, ``SP0``]: term array
val svm = (2, <Redblackmap(1)>): int * (term, term * int) SVM.dict
val tmpname = "/var/tmp//MLTEMPECqLFC": string
val vc = 1: int
```
Then, the file contents of `in_name` ("/var/tmp//MLTEMPECqLFC.cnf") is the following:
```
c File /var/tmp//MLTEMPECqLFC.cnf generated by HolSatLib
c
p cnf 1 1
1 
0
```

But the above DIMACS is *wrong*. Here the input term of `SAT_ORACLE` is `b ==> T`, which is equivalent to `^b \/ T`. Recall that the goal was to check if this term is a tautology (`!b. b ==> T`) and the method is to let SAT solver to check the negation of the input term, i.e. `~(b ==> T)`, which is `b /\ F`, a CNF.

In DIMACS format, the line `p cnf n m` indicates there are totally n Boolean variables and m clauses, and the rest lines each represent one clause, each is a disjunction of involved variables (zero or more), *ended* with zero. Thus `b` is `1 0`, and `F` should be just `0`. The above DIMACS file generated by HOL has two issues:

1. The line `p cnf ...` has wrong numbers. It should be `p cnf 1 2` instead of `p cnf 1 1` (line 3).
2. Somehow the line for the first clause (`1`) misses the ending zero, i.e., it should be `1 0` (line 4).

There's a MiniSat web interface [1] taking DIMACS inputs. If I put the above DIMACS contents as is, the solving result is SAT. But after manual fixes according to the above two issues (only the 2nd is fatal), the result is UNSAT, the correct one.

[1] http://logicrunch.it.uu.se:4096/~wv/minisat/
